flink-runtime/src/main/java/org/apache/flink/runtime/security/FlinkUserSecurityManager.java

```diff
@@ -0,0 +1,165 @@
+ * against user mistakes by gracefully handling them informing users rather than causing silent unavailability.
+ */
+public class FlinkUserSecurityManager extends SecurityManager {
+
+	/**
+	 * The mode of how to handle user code attempting to exit JVM.
+	 */
+	public enum CheckExitMode {
+		/** No check is enabled, that is allowing exit without any action. */
+		DISABLED,
+		/** Warn by logging but still allowing exit to be performed. */
+		WARN,
+		/** Throw exception when exit is attempted disallowing JVM termination. */
+		THROW,
+	}
+
+	static final Logger LOG = LoggerFactory.getLogger(FlinkUserSecurityManager.class);
+
+	private final SecurityManager originalSecurityManager = System.getSecurityManager();
+	private ThreadLocal<Boolean> monitorSystemExit = new InheritableThreadLocal<>();
+	private CheckExitMode checkExitMode;
+
+	public FlinkUserSecurityManager(CheckExitMode checkExitMode) {
+		super();
+		this.checkExitMode = checkExitMode;
+
+		LOG.info("FlinkUserSecurityManager is created with {} system exit check (previous security manager is {})",
+			this.checkExitMode, originalSecurityManager != null ? originalSecurityManager : "not existing");
+	}
+
+	/**
+	 * Instantiate FlinkUserSecurityManager from configuration. Return null if no security manager
+	 * check is needed, so that a caller can skip setting security manager avoiding runtime check cost,
+	 * if there is no security check set up already. Use {@link #setFromConfiguration} helper, which
+	 * handles disabled case.
+	 *
+	 * @param configuration to instantiate the security manager from
+	 *
+	 * @return FlinkUserSecurityManager instantiated baesd on configuration. Return null if disabled.
+	 */
+	public static FlinkUserSecurityManager fromConfiguration(Configuration configuration) {
+		final String checkExitModeConfig = configuration.getString(SecurityOptions.CHECK_SYSTEM_EXIT);
+		final CheckExitMode checkExitMode;
+
+		try {
+			checkExitMode = CheckExitMode.valueOf(checkExitModeConfig.toUpperCase());
+		} catch (Exception ex) {
+			throw new IllegalConfigurationException(
+				String.format("%s is invalid configuration for %s.", checkExitModeConfig, SecurityOptions.CHECK_SYSTEM_EXIT.key()),
+				ex);
+		}
+
+		// If no check is enabled, return null so that caller can avoid setting security manager not to incur any runtime cost.
+		if (checkExitMode == CheckExitMode.DISABLED) {
+			return null;
+		}
+		// Add more configuration parameters that need user security manager (currently only for system exit).
+		return new FlinkUserSecurityManager(checkExitMode);
+	}
+
+	public static void setFromConfiguration(Configuration configuration) {
+		final FlinkUserSecurityManager flinkUserSecurityManager =
+			FlinkUserSecurityManager.fromConfiguration(configuration);
+		if (flinkUserSecurityManager != null) {
+			System.setSecurityManager(flinkUserSecurityManager);
```

Comment:
We might want to add more error handling here as well: Compare with `ExitTrappingSecurityManager.replaceGracefulExitWithHaltIfConfigured(..)`